### ISSUE TITLE
Semantic tokens should be refreshed quickly, even in large files

### ISSUE THREAD

**Issue by mainej** [clojure-lsp/clojure-lsp#970]
**Describe the bug**

After dragging a clause in a large file, it takes an excessively long time to fix semantic tokens. At the moment for me in clojure.core it's taking 25s between the drag and the time that the tokens are fixed. The time seems to be getting longer the more I play with it. 

**To Reproduce**
1. Go to the definition of `defmacro`, in clojure.core.
2. Drag clauses within the let form
3. Observe that it can take seconds or tens of seconds for the semantic token coloring to correct itself.

What's particularly strange is that the logs show that clojure-lsp is actually returning the semantic token data pretty quickly. After the edit, there's a 300ms debounce time, followed by 1500-2000ms of clj-kondo analysis time, and then 30ms later the tokens are returned. Let's say 2 seconds on average. 

However, watching the CPU tells a different story. clojure-lsp uses an entire core for tens of seconds after the edit. The editor, even though it's been given the fixed semantic token data, doesn't make any changes until clojure-lsp is done doing whatever it's doing. Unfortunately, clojure-lsp doesn't emit any logs at the end of its CPU hogging, so I don't know what it's been doing.

This was noticed in Calva, but, the experience in Emacs is similar. Long pauses, getting longer the more I play with it.

**Expected behavior**
The semantic tokens are fixed soon after the file is analyzed.

**Comment by ericdallo** [comment]
Hum, so the tokens calculations are fast but even so it takes time? I think it's related with multiple changes where we would wait for tokens because of kondo analysis running on the file?

**Comment by mainej** [comment]
> wait for tokens because of kondo analysis running on the file

That can be a reason for pauses, but I don't think it's what's happening in this case... Those tools that wait for analysis to finish always print something to the logs when they're done, and in this case, nothing gets printed at the end of the pause. Here's a more complete sequence:

1. client requests drag
2. server calculates drag edits, and returns them
3. client applies drag edits (i.e. user sees the buffer change)
4. server sends cursor positioning
5. client positions cursor (i.e. user sees the cursor move)
6. client sends did-change (I think... something triggers re-analysis)
7. server starts re-analysis (after a debounce period)
8. client requests semantic tokens; the server delays its response until re-analysis is finished.
9. server finishes re-analysis
10. server sends diagnostics, and the client applies them (i.e., user sees squiggles added, removed, or moved)
11. server sends semantic tokens, but the client does not apply them

At this point, about 2 seconds have passed. After this point, the server doesn't create any more log lines. It thinks it has delivered the semantic tokens, but the semantic token colors are still wrong in the client.

12. tens of seconds pass; 100% of a core is being used.
13. the CPU usage drops and the client fixes the semantic tokens (i.e., the user sees the colors change)

Strange. What is the server doing during this time? My first guess was notifying references. But that's not it... the server logs that it's able to notify references before the long pause starts, taking only 39ms.

Also, what is the client doing? If it has the semantic tokens, why isn't it applying them?

To further confuse things, Emacs and Calva exhibit similar but slightly different behavior. Emacs takes about 10 seconds to fix the token colors, while Calva takes about 30 seconds. And while Emacs is waiting, the process that is taking a lot of CPU is `java`, but while Calva is waiting, the process is `clojure-lsp`. There's a chance this is actually two separate problems with similar symptoms.

Anyway I'm not sure what's going on... Needs some tracing.

**Comment by ericdallo** [comment]
Are you sure that core process is being used on clojure-lsp? maybe it's emacs (single threaded) computing tokens?

**Comment by ericdallo** [comment]
Anyway, it'd be nice to know what server is doing if it's its fault indeed, those tens of seconds are too much indeed, we need to dig it and confirm it's server's fault

**Comment by mainej** [comment]
I'm really not sure who's to blame. I've figured out that when Calva starts the language server, it shows up in my process list as `clojure-lsp`, but when Emacs starts it, it shows up as `java`. So probably, it's the same problem in both cases.

You're right though: there are two possibly unrelated questions:
1. Why is the server using so much CPU?
2. If the client has the token data early, why does it take so long to apply it? Does it just take a long time to apply tokens in large files?

If they're related, then understanding and fixing the server CPU usage will also fix the client. If they're not, then fixing the server CPU usage won't help the client. My suspicion is that they're related, because the token colors update fractions of a second after the CPU usage drops, but I don't have any hard proof.

**Comment by ericdallo** [comment]
yes, makes sense to be server's fault, it's curious though, the semantic tokens logic is just a bunch of conversion to vectors of ints which is lighter to return to client, so a performance issue would be probably iterating all analysis of that file which should not be that bad in performance, I suspect it's related with we blocking something waiting for analysis to complete, still those tens of seconds are too much, maybe some bug on that logic

**Comment by mainej** [comment]
I really don't think it has anything to do with the server's part of semantic tokens, either calculating them or waiting to deliver them... Immediately after analysis completes the server logs that it has stopped waiting, calculated the tokens, and sent them to the client, with about 2000ms of wait time and 30ms of calculation time. That all happens before the long pause starts.

I also don't think that anything else is blocking on analysis. If there were something blocking on analysis, I'd expect two log lines after the long pause: 1. a log line reporting how long it took to run analysis, and 2. a log line reporting that the thing that was blocked resumed and finished its task. But we don't see either of those log lines.

That said, it feels like something is blocking somewhere.

**Comment by ericdallo** [comment]
that's bizarre, I think we will need to use jconsole or something to monitor the threads... oh, can flamegraph help with that? I don't think so since we need to wrap it in some function which we don't know what is causing the issue

**Comment by mainej** [comment]
Yeah, my plan is to use `clj-async-profiler`. It has two modes—you can wrap a block that you want it to profile, or you can start it and later stop it. I'm not exactly sure how to set it up, but I plan to use the second mode... start the profiler, execute the drag, wait for the long pause, stop the profiler, and review the flamegraph.

**Comment by mainej** [comment]
I got a flamegraph, and I have a theory. I was seeing this in the log:

```
2022-05-05T23:01:34.068Z  INFO [clojure-lsp.feature.file-management:181] - changes analyzed by clj-kondo took 1437ms
2022-05-05T23:01:34.070Z  INFO [clojure-lsp.server:251] - :analyze-file 1438ms
...
2022-05-05T23:01:34.138Z  INFO [clojure-lsp.handlers:441] - :semantic-tokens-range 13ms - waited 1557ms
2022-05-05T23:01:34.215Z  INFO [clojure-lsp.handlers:434] - :semantic-tokens-full 31ms - waited 1572ms
```

My interpretation was two-fold: A) the full semantic tokens took 31ms to calculate, and B) the system had delivered the tokens to the client. Both of these are false.

Instead, we're a victim of laziness. What has actually happened when we see these log lines, is that the code has calculated a lazy seq that will eventually contains all the semantic tokens. Realizing that lazy seq (which happens in the coerecer, after the log line has been emitted by the handler) takes tens of seconds.

So, there are two things I'd like to do.
1. I'll add a `doall` in the handler, so that the logs report the time to calculate all the tokens, not the time to prepare to calculate all the tokens.
2. See if I can make the semantic tokens code faster. The flamegraph suggests that the huge majority of the time is from calling `nth`. It looks like we're doing that in a loop, and the thing we're calling it on is a list, which would be an exponential algorithm (as opposed to a linear algorithm if we were calling it on a vector). There's an easy and idiomatic way to convert that to a linear algorithm, without needing either a list or a vector, so I'll give that a shot.

**Comment by mainej** [comment]
I added the `doall`. It confirms that `:semantic-tokens-full` is what's taking so long. Now to see if I can make it faster.

```
2022-05-05T23:31:27.886Z  INFO [clojure-lsp.feature.file-management:181] - changes analyzed by clj-kondo took 1713ms
2022-05-05T23:31:27.888Z  INFO [clojure-lsp.server:251] - :analyze-file 1715ms
...
2022-05-05T23:31:27.984Z  INFO [clojure-lsp.handlers:441] - :semantic-tokens-range 19ms - waited 1981ms
2022-05-05T23:31:35.304Z  INFO [clojure-lsp.handlers:434] - :semantic-tokens-full 7256ms - waited 1984ms
```

**Comment by mainej** [comment]
And also confirmed, converting to a linear algorithm is faster; for clojure.core about 96% faster:

```
2022-05-05T23:40:35.895Z  INFO [clojure-lsp.feature.file-management:181] - changes analyzed by clj-kondo took 2064ms
2022-05-05T23:40:35.901Z  INFO [clojure-lsp.server:251] - :analyze-file 2070ms
...
2022-05-05T23:40:36.245Z  INFO [clojure-lsp.handlers:434] - :semantic-tokens-full 313ms - waited 2400ms
```

I'm going to play a little with transducers to see if I can make it slightly faster.

**Comment by mainej** [comment]
Abandoning transducers... there are too many non-transducer functions interleaved in the algorithm.

**Comment by mainej** [comment]
Compared to 7000ms, 300ms is great, but it's hard to be really enthusiastic... I _thought_ this was taking 30ms. Maybe there's more to be optimized in semantic tokens, but this is a good start.

The experience in the editors is better. You make the drag, wait a few seconds, then the diagnostics and semantic tokens come in together. If you pay close attention, the diagnostics come in a fraction of a second before the tokens. A few seconds is still a long wait, but it's OK because semantic tokens and diagnostics are more informative than functional.

Most importantly, it's nice to reduce CPU usage. The CPU spike when you edit a large file is getting much smaller.

**Comment by mainej** [comment]
For future reference... here's the profiling code I used.

```clojure
(comment
  (require '[clojure-lsp.debug :as debug]
           '[clj-async-profiler.core :as profiler])
  (profiler/start {})

  (def r (profiler/stop {:min-width 5
                         :return-file true}))
  (debug/open-file r))
```

I put this in `clojure.core`, and connected to the clojure-lsp nREPL. I evaluated the require and the `profiler/start` line. Then I navigated the cursor to the place I wanted to drag and did the drag. I waited until the tokens came in, then went back and evaluated the `(def r (profiler/stop...))` form. Then I ran `(debug/open-file r)` and was able to see the flamegraph. It was immediately obvious that 80% of the time was in `nth`, with most of the rest of the time being in clj-kondo. I could see `nth` was related to semantic tokens because `nth` was called by `clojure-lsp.feature.semantic-tokens/absolute-token->relative-token`. But it took me a minute to figure out why `clojure-lsp.feature.semantic-tokens/full-tokens` wasn't in the call stack. Eventually I noticed that `coercer` was in the call stack and worked out that the bulk of the time was probably spent realizing a lazy-seq returned by `full-tokens`. All-in-all, pretty easy debugging.
